tiledb/sm/array_schema/array_schema.h

```diff
@@ -73,6 +73,47 @@ class ArraySchema {
   /** Constructor. */
   ArraySchema(ArrayType array_type);
 
+  /** Constructor.
+   * @param allows_dups True if the (sparse) array allows coordinate duplicates.
+   * @param array_uri An array name attached to the schema object.
+   * @param array_type The array type.
+   * @param attribute_map Maps each name to the corresponding attribute object.
+   * @param attributes The array attributes.
+   * @param capacity The tile capacity for the case of sparse fragments.
+   * @param cell_order The cell order.
+   * @param cell_var_offsets_filters
+   *    The filter pipeline run on offset tiles for var-length attributes.
+   * @param cell_validity_filters
+   *    The filter pipeline run on validity tiles for nullable attributes.
+   * @param coords_filters The filter pipeline run on coordinate tiles.
+   * @param dim_map Maps each name to the corresponding dimension object.
+   * @param domain The array domain.
+   * @param tile_order The tile order.
+   * @param version The format version of this array schema.
+   * @param timestamp_range The timestamp the array schema was written.
+   * @param uri The URI of the array schema file.
+   * @param name
+   *    The file name of array schema in the format of timestamp_timestamp_uuid.
+   **/
```

Comment:
The order of all these parameters is quite confusing. They can be grouped, I think, into four categories:
1. The array as a stored entity. Primarily this `array_uri`, but also the format version and some others.
2. The semantics of the array as a whole. This is anything that doesn't fit into the domain (coordinates) or codomain (attributes).
3. The domain. Not just the `Domain` object itself, but also `tile_order` (which should be part of the domain, but whatever for now).
4. The codomain. (There really should be a separate class for this, but again whatever for now.) Everything about attributes, their filters, etc.

I think that some of these arguments could be eliminated by computing them in the constructor. `dim_map` IIRC is simply taken from the domain.